_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+* TRACE: Allows for a loop-back test with debug information. This method is not included in default authorization checking, and should be disabled in production, as it can be a security risk.
+
+The book concludes its elaboration on the second level of Richardson’s Maturity with an overview of some of the more popular technologies and frameworks and how they tackle the creation of a consumer for the services defined with REST. 
+These are rather straightforward for those developers that routinely use them, so we will not go deeper into them.
+
+## Level 3 - Hypermedia Controls
+
+Hypermedia systems extend the resource state that is being manipulated by the services with additional characteristics. 
+The resource state becomes a combination of:
+* The values of the individual variables that make up the resource
+* Links to related resources
+* Links to manipulate the current resource (creating, updating, deleting…)
+* Evaluation results of business rules encompassing the resource and other related resources
+
+The links mentioned make up a domain application protocol (DAP) that advertises all possible interactions with the resource. 
+The consumers of this service use these published interactions to discover how to interact with the resource. 
+Hyperlinks are the weapon of choice when adding such links to interactions. 
+A DAP also consists of two other key components: Media Types (as mentioned before), and HTTP Idioms (which make up the HTTP uniform interface: verbs, standard headers, error codes…). 
+In short: Link relation values tell why the consumer should activate a hypermedia control by stating the role of the linked resource, the media type tells the consumer what it can expect as the response of a link, and the idioms manipulate the resources represented by the links.
+
+The resource state can be in any format as REST’s hypermedia tenet does not force any set media type. 
+IANA is the official registry of MIME media types and maintains a [list of all the universally accepted MIME types](http://www.iana.org/assignments/media-types/media-types.xhtml){:target="_blank" rel="noopener noreferrer"}.
+The resource should indicate in the Content-Type header of its responses the ideal way of interpreting it. 
+Similarly, the Accept header is part of the HTTP spec for requests. 
+In this header the consumer can indicate which media type it wishes to receive. 
+Based on these two headers, a system of content negotiation is set up where the request of a consumer can be matched to the proper implementation to return the desired result and how it is conveyed.
+
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2021-01-04-Designing-REST-services/mediatypes.png' | prepend: site.baseurl }}" alt="Workspace" class="image" style="margin:0px auto; max-width:100%">
```

Comment:
If possible I'd use a solid background for the image instead of a transparent one.
In dark mode it is not readable:
![image](https://user-images.githubusercontent.com/1644089/103444759-09d40180-4c6c-11eb-88fd-ab9d3697050d.png)

Reply:
done